File: src/GeoData/Distance.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Simple.GeoData
{
    static class Distance
    {
        public static double haversineDistance(Point from, Point to)
        {
            double r = 6365000;
            double lat1 = from[1] * Math.PI / 180;
            double lat2 = to[1] * Math.PI / 180;
            double lon1 = from[0] * Math.PI / 180;
            double lon2 = to[0] * Math.PI / 180;
            double a = Math.Pow(Math.Sin((lat2 - lat1) / 2), 2);
            double b = Math.Pow(Math.Sin((lon2 - lon1) / 2), 2);
            return 2 * r * Math.Asin(Math.Sqrt(a + Math.Cos(lat1) * Math.Cos(lat2) * b));
        }

        public static double euclideanDistance(Point a, Point b)
        {
            return Math.Sqrt(Math.Pow(a[0] - b[0], 2) + Math.Pow(a[1] - b[1], 2));
        }
    }
}

File: src/go-routing/geo/distance.go
package geo

import (
	"math"
)

func HaversineDistance(from, to Coord) float64 {
	r := 6365000.0
	lat1 := from[1] * math.Pi / 180
	lat2 := to[1] * math.Pi / 180
	lon1 := from[0] * math.Pi / 180
	lon2 := to[0] * math.Pi / 180
	a := math.Pow(math.Sin(float64((lat2-lat1)/2)), 2)
	b := math.Pow(math.Sin(float64((lon2-lon1)/2)), 2)
	return 2.0 * r * math.Asin(math.Sqrt(a+math.Cos(float64(lat1))*math.Cos(float64(lat2))*b))
}

func HaversineLength(points CoordArray) float64 {
	r := 6365000.0
	length := float64(0)
	for i := 0; i < len(points)-1; i++ {
		lat1 := points[i][1] * math.Pi / 180
		lat2 := points[i+1][1] * math.Pi / 180
		lon1 := points[i][0] * math.Pi / 180
		lon2 := points[i+1][0] * math.Pi / 180
		a := math.Pow(math.Sin(float64(lat2-lat1)/2), 2)
		b := math.Pow(math.Sin(float64(lon2-lon1)/2), 2)
		length += 2 * r * math.Asin(math.Sqrt(a+math.Cos(float64(lat1))*math.Cos(float64(lat2))*b))
	}
	return length
}

func EuclideanDistance(a, b Coord) float64 {
	d_lon := float64(a[0]) - float64(b[0])
	d_lat := float64(a[1]) - float64(b[1])
	return math.Sqrt(math.Pow(d_lon, 2) + math.Pow(d_lat, 2))
}
